tests/unit/s2n_connection_serialize_test.c

```diff
@@ -473,6 +490,271 @@ int main(int argc, char **argv)
+
+            /* Client sets up a resumption connection with the received session ticket data */
+            size_t cb_session_data_len = s2n_stuffer_data_available(&cb_session_data);
+            EXPECT_SUCCESS(s2n_connection_set_session(new_client_conn, cb_session_data.blob.data, cb_session_data_len));
+            EXPECT_SUCCESS(s2n_stuffer_rewrite(&cb_session_data));
+
+            /* Negotiate successful resumption connection */
+            EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, new_client_conn));
+            EXPECT_FALSE(IS_FULL_HANDSHAKE(new_client_conn));
+            EXPECT_FALSE(IS_FULL_HANDSHAKE(server_conn));
+        };
+
+        /* Server is serialized. Can write a session ticket after deserialization. */
+        {
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+
+            /* Wipe ticket from previous test */
+            EXPECT_SUCCESS(s2n_stuffer_wipe(&cb_session_data));
+
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, resumption_config));
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, resumption_config));
+
+            DEFER_CLEANUP(struct s2n_test_io_stuffer_pair io_pair = { 0 }, s2n_io_stuffer_pair_free);
+            EXPECT_OK(s2n_io_stuffer_pair_init(&io_pair));
+            EXPECT_OK(s2n_connections_set_io_stuffer_pair(client_conn, server_conn, &io_pair));
+
+            EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server_conn), S2N_TLS13);
+
+            /* Get the client to do a read to pick up the session ticket. We won't use this one;
+             * we're interested in the tickets the server sends after the transfer. */
+            EXPECT_OK(s2n_send_and_recv_test(server_conn, client_conn));
+            EXPECT_SUCCESS(s2n_stuffer_wipe(&cb_session_data));
+
+            /* Serialize the server */
+            uint8_t buffer[S2N_SERIALIZED_CONN_TLS12_SIZE] = { 0 };
+            EXPECT_SUCCESS(s2n_connection_serialize(server_conn, buffer, sizeof(buffer)));
+
+            /* Initialize new server connection and deserialize session */
+            DEFER_CLEANUP(struct s2n_connection *new_server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(new_server_conn);
+            EXPECT_SUCCESS(s2n_connection_deserialize(new_server_conn, buffer, sizeof(buffer)));
+
+            /* Re-initialize IO pipes */
+            EXPECT_OK(s2n_connections_set_io_stuffer_pair(client_conn, new_server_conn, &io_pair));
```

Comment:
You're reusing the possibly not empty IO stuffers here too. Can you be more careful with keeping the IO clean? Maybe use contexts + DEFER_CLEANUP? That might also break up the test nicely.

You want to make sure you're REALLY transferring, and not relying on anything from the original connection. I wouldn't even wipe the connections. New everything.